Task 4

dhorvath333 opened the pull request:
- [x] Task 4.1: RDS database created with PostgreSQL with "product" and "stock" tables
- [x] Task 4.2: Lambda functions are connected and fetching data from database
- [x] Task 4.3: POST/products lambda function created (see link below)
- [x] FE: FE is integrated with the product-service and fetches products from the database (see FE link below)
- [x] +1: POST /products returns 400 on invalid data (e.g.: if price is <=0)
- [x] +1: Lambda functions return 500 on DB connection or query error
- [x] +1: Lambda functions console log for each request with the sent data in case of POST /products
- [x] +1: POST /products insert into query implemented with transactions


---------------------------------------------------------------------------
FE link:
https://d3i6ehyl8qm7qy.cloudfront.net/

Get product list API (GET /products):
https://fdu9dsnyeb.execute-api.eu-central-1.amazonaws.com/dev/products

Create products API (POST /products):
https://fdu9dsnyeb.execute-api.eu-central-1.amazonaws.com/dev/products
POST body:
```
{
    "title": "Test product",
    "description": "Test product description",
    "price": 123,
    "count": 5
}
```

nodm commented on product-service/src/functions/product-by-id/handler.ts:
```diff
@@ -0,0 +1,50 @@
+  try {
+    const db = getDataBase();
+
+    const query = await db.query('SELECT p.*, count FROM product p LEFT JOIN stock on product_id = p.id WHERE p.id = $1', [productId]);
```
Great job!
Just a quick tip: the error is thrown while comparing uuid and string.
You can check it [here](https://fdu9dsnyeb.execute-api.eu-central-1.amazonaws.com/dev/products/271189da-53c3-4c90-af71-71bb4958c8e41)
So the quick fix is `WHERE p.id::text = $1`

dhorvath333 replied:
Thanks! fixed

nodm commented on product-service/src/functions/create-product/handler.ts:
```diff
@@ -0,0 +1,79 @@
+  const isTitleValid = validateTitle(title);
+  const isDescriptionValid = validateDescription(description);
+  const isPriceValid = validatePrice(price);
+  const isCountValid = validateCount(count);
```
Good job, but using a schema validator like [Joi](https://joi.dev/) in the production code is better.

dhorvath333 replied:
Yes, true. I added serverless schema validation just in case, but leave this part as it is

VladislavLobakh commented on product-service/src/functions/create-product/handler.ts:
```diff
@@ -0,0 +1,79 @@
+export const createProduct = async (event: Partial<APIGatewayProxyEvent>): Promise<APIGatewayProxyResult> => {
+  const { title, description, price, count } = (event.body as any) as Product;
+
+  console.log('POST /products', { title, description, price, count });
```
I guess all event params should be presented here, not only body

VladislavLobakh replied:
pathParameters,
        queryStringParameters,
        body

dhorvath333 replied:
That's up to interpretation, I only logged the data that matters for the app here. But fixed it in a commit and now it logs the whole event object

VladislavLobakh commented on product-service/src/functions/create-product/handler.ts:
```diff
@@ -0,0 +1,79 @@
+      const product = createProductQuery.rows[0];
+      await db.query('INSERT INTO stock (product_id, count) values ($1, $2)', [product.id, count]);
+      await db.query('COMMIT')
+      await db.end();
```
must be ended in case of failure as well, so move it to finally section

dhorvath333 replied:
Good point! fixed

VladislavLobakh commented on product-service/src/functions/product-list/handler.ts:
```diff
@@ -0,0 +1,35 @@
+
+    const query = await db.query('SELECT p.*, count FROM product p LEFT JOIN stock on product_id = p.id');
+    const products = query?.rows;
+    await db.end();
```
the same, move to finally section

dhorvath333 replied:
fixed

VladislavLobakh commented:
https://fdu9dsnyeb.execute-api.eu-central-1.amazonaws.com/dev/products/aaaa
wrong response code

VladislavLobakh commented:
https://d3i6ehyl8qm7qy.cloudfront.net/admin/products/c374ad06-e48e-4641-89c9-6fc4ea599dd5
no product data presented, API is not integrated

VladislavLobakh commented:
Create product API is not integrated with FE app
<img width="1567" alt="Screenshot 2022-07-27 at 21 27 05" src="https://user-images.githubusercontent.com/24480136/181345572-3454a916-1c07-4256-877d-5e36c4714d0e.png">

VladislavLobakh commented on product-service/serverless.ts:
```diff
@@ -0,0 +1,49 @@
+    },
+    environment: {
+      AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
+      NODE_OPTIONS: '--enable-source-maps --stack-trace-limit=1000',
```
Extend your serverless.yml file with credentials to your database instance and pass it to lambda’s environment variables section.

dhorvath333 replied:
But it also says not to commit those environment variables in the PR

VladislavLobakh commented:
Validation is working not properly, there's possibility to send such request. Types checks should be done
{
  "title": "Test product 3 (review)",
  "description": 300,
  "price": 1,
  "count": "400",
  "aaa": "aaaa"
}

dhorvath333 commented:
> https://fdu9dsnyeb.execute-api.eu-central-1.amazonaws.com/dev/products/aaaa wrong response code

fixed

> https://d3i6ehyl8qm7qy.cloudfront.net/admin/products/c374ad06-e48e-4641-89c9-6fc4ea599dd5 no product data presented in FE app, API is not integrated

fixed


> Create product API is not integrated with FE app <img alt="Screenshot 2022-07-27 at 21 27 05" width="1567" src="https://user-images.githubusercontent.com/24480136/181345572-3454a916-1c07-4256-877d-5e36c4714d0e.png">

fixed
